File: benchmark/c/primes.c
/*
 * primes.c - Prime Sieve Benchmark for C
 * Tests memory allocation and CPU-bound computation
 */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>

/* Sieve of Eratosthenes */
int sieve_of_eratosthenes(int limit) {
    bool *is_prime = malloc(limit + 1);
    if (!is_prime) {
        fprintf(stderr, "Failed to allocate memory\n");
        exit(1);
    }

    /* Initialize all as potentially prime */
    for (int i = 0; i <= limit; i++) {
        is_prime[i] = true;
    }
    is_prime[0] = is_prime[1] = false;

    /* Sieve */
    for (int i = 2; i * i <= limit; i++) {
        if (is_prime[i]) {
            for (int j = i * i; j <= limit; j += i) {
                is_prime[j] = false;
            }
        }
    }

    /* Count primes */
    int count = 0;
    for (int i = 2; i <= limit; i++) {
        if (is_prime[i]) count++;
    }

    free(is_prime);
    return count;
}

/* Get current time in milliseconds */
long get_time_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec * 1000 + ts.tv_nsec / 1000000;
}

int main(void) {
    int limit = 1000000;
    long start, end;

    start = get_time_ms();
    int count = sieve_of_eratosthenes(limit);
    end = get_time_ms();

    printf("Primes up to %d: %d\n", limit, count);
    printf("Sieve time: %ldms\n", end - start);

    return 0;
}

File: benchmark/java/Primes.java
// Primes.java - Prime Sieve Benchmark for Java
// Tests memory allocation and CPU-bound computation

public class Primes {
    // Sieve of Eratosthenes
    public static int sieveOfEratosthenes(int limit) {
        boolean[] isPrime = new boolean[limit + 1];
        for (int i = 0; i <= limit; i++) {
            isPrime[i] = true;
        }
        isPrime[0] = false;
        isPrime[1] = false;

        for (int i = 2; i * i <= limit; i++) {
            if (isPrime[i]) {
                for (int j = i * i; j <= limit; j += i) {
                    isPrime[j] = false;
                }
            }
        }

        int count = 0;
        for (int i = 2; i <= limit; i++) {
            if (isPrime[i]) count++;
        }
        return count;
    }

    public static void main(String[] args) {
        int limit = 1000000;

        long start = System.currentTimeMillis();
        int count = sieveOfEratosthenes(limit);
        long end = System.currentTimeMillis();

        System.out.println("Primes up to " + limit + ": " + count);
        System.out.println("Sieve time: " + (end - start) + "ms");
    }
}
